chain/network/src/accounts_data/mod.rs

```diff
@@ -0,0 +1,240 @@
+            epochs.0.entry(e.id.clone()).or_insert_with(|| {
+                has_new = true;
+                Epoch::new(&e.priority_accounts)
+            });
+        }
+        has_new
+    }
+
+    /// Selects new data and verifies the signatures.
+    /// Returns the verified new data and an optional error.
+    /// Note that even if error has been returned the partially validated output is returned
+    /// anyway.
+    fn verify(&self, data: Vec<SignedAccountData>) -> (Vec<SignedAccountData>, Option<Error>) {
+        // Filter out non-interesting data, so that we never check signatures for valid non-interesting data.
+        // Bad peers may force us to check signatures for fake data anyway, but we will ban them after first invalid signature.
+        // It locks epochs for reading for a short period.
+        let mut found = HashSet::new();
+        let mut data_and_keys = vec![];
+        let epochs = self.epochs.read();
+        for d in data {
+            // We want the communication needed for broadcasting per-account data to be minimal.
+            // Therefore broadcasting multiple datasets per account is considered malicious
+            // behavior, since all but one are obviously outdated.
+            let data_id = (d.epoch_id.clone(), d.account_id.clone());
+            if found.contains(&data_id) {
+                return (vec![], Some(Error::SingleAccountMultipleData));
+            }
+            found.insert(data_id);
+            // There is a limit on the amount of RAM occupied by per-account datasets.
+            // Broadcasting larger datasets is considered malicious behavior.
+            if d.payload().len() > network_protocol::MAX_ACCOUNT_DATA_SIZE_BYTES {
+                return (vec![], Some(Error::DataTooLarge));
+            }
+            // It is fine to broadcast account data for epochs that we don't care about.
+            let epoch = match epochs.0.get(&d.epoch_id) {
+                Some(e) => e,
+                None => continue,
+            };
+            // It is NOT fine to broadcast account data which is not relevant to the given epoch.
+            let account = match epoch.0.get(&d.account_id) {
+                Some(a) => a,
+                None => return (vec![], Some(Error::InvalidAccount)),
+            };
+            // It is fine to broadcast data we already know about.
+            if !account.is_new(d.timestamp) {
+                continue;
+            }
+            data_and_keys.push((d, account.key.clone()));
+        }
+        drop(epochs);
+
+        // We verify signatures synchronously for now.
+        // To verify signatures in parallel, we should have a way to stop verification at the first invalid one.
+        let mut data = vec![];
+        for (d, key) in data_and_keys {
+            if !d.payload().verify(&key) {
+                return (data, Some(Error::InvalidSignature));
+            }
+            data.push(d);
+        }
+        return (data, None);
```

Comment:
```suggestion
        (data, None)
```

I feel like I make redundant return suggestion fairly frequently, please try to stick to idiomatic rust :)

Reply:
I'm not doing that on purpose, and I think that usually I get it right (just not always). The difference is marginal, so I need to really concentrate to spot things like that. If that's idiomatic, maybe "cargo fmt" should do those fixes?

Reply:
`cargo fmt` doesn't do this as this feature needs a bit of control-flow analysis to detect expression in tail positions. But I believe `cargo clippy` would complain, there's an on-by-default lint for that: 

https://rust-lang.github.io/rust-clippy/master/#needless_return